crates/adapters/src/secret_resolver.rs

```diff
@@ -0,0 +1,387 @@
+    let connector_config_resolved = serde_yaml::from_value(yaml_value_resolved)
+        .map_err(|_e| SecretRefResolutionError::DeserializationFailed)?;
+    Ok(connector_config_resolved)
+}
+
+/// Resolves recursively the secret references in the YAML.
+fn resolve_secret_references_in_yaml(
+    secrets_dir: &Path,
+    value: Value,
+) -> Result<Value, SecretRefResolutionError> {
+    Ok(match value {
+        Value::Null => Value::Null,
+        Value::Bool(b) => Value::Bool(b),
+        Value::Number(n) => Value::Number(n),
+        Value::String(s) => {
+            Value::String(resolve_potential_secret_reference_string(secrets_dir, s)?)
+        }
+        Value::Sequence(seq) => Value::Sequence(
+            seq.into_iter()
+                .map(|v| resolve_secret_references_in_yaml(secrets_dir, v))
+                .collect::<Result<Vec<Value>, SecretRefResolutionError>>()?,
+        ),
+        Value::Mapping(mapping) => {
+            let mut new_mapping = Mapping::new();
+            for (k, v) in mapping.into_iter() {
+                if let Some(_existing) =
+                    new_mapping.insert(k, resolve_secret_references_in_yaml(secrets_dir, v)?)
+                {
+                    return Err(SecretRefResolutionError::DuplicateKeyInMapping);
+                }
+            }
+            Value::Mapping(new_mapping)
+        }
+        Value::Tagged(mut tag_val) => {
+            tag_val.value = resolve_secret_references_in_yaml(secrets_dir, tag_val.value)?;
+            Value::Tagged(tag_val)
+        }
+    })
+}
+
+/// Resolves a string which can potentially be a secret reference.
+fn resolve_potential_secret_reference_string(
+    secrets_dir: &Path,
+    s: String,
+) -> Result<String, SecretRefResolutionError> {
+    match MaybeSecretRef::new(s) {
+        Ok(maybe_secret_ref) => match maybe_secret_ref {
+            MaybeSecretRef::String(plain_str) => Ok(plain_str),
+            MaybeSecretRef::SecretRef(secret_ref) => match &secret_ref {
+                SecretRef::Kubernetes { name, data_key } => {
+                    // Secret reference: `${secret:kubernetes:<name>/<data key>}`
+                    // File location: `<secrets dir>/kubernetes/<name>/<data key>`
+                    let path = Path::new(secrets_dir)
+                        .join("kubernetes")
+                        .join(name)
+                        .join(data_key);
+
+                    // If the file does not exist or is not a file, produce a custom error informing
+                    // of this fact rather than generically stating we are unable to read.
+                    if path.is_file() {
+                        match fs::read_to_string(&path) {
+                            Ok(content) => Ok(content),
+                            Err(_e) => {
+                                // The read error is not part of the error returned to prevent
+                                // displaying any of the secret content
```

Comment:
I don't think fs::read_to_string puts anything it reads into an error. (If that's a problem then ErrorKind would definitely not).

Reply:
True, I can limit it to ErrorKind to be on the cautious side but still provide some info.